### PR TITLE
[REEF-1639] Refactor RunnableProcess and move State logic into a separate enum

### PR DESCRIPTION
**Pull request by motus**
JIRA: [REEF-1639](https://issues.apache.org/jira/browse/REEF-1639)

### CODE REVIEW

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/process/RunnableProcessState.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  ENDED;
+
+  /**
+   * Check whether a transition from current state to the given one is legal.
+   * @param toState destination state.
+   * @return True if the state transition is legal, false otherwise.
+   */
+  public boolean isLegal(final RunnableProcessState toState) {
+
+    switch (this) {
```
Comment (tcNickolas) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/process/RunnableProcessState.java]: Why is transition from `INIT` to `INIT` legal, but transitions from `RUNNING` to `RUNNING` and from `ENDED` to `ENDED` illegal? Seems suspicious to me

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/process/RunnableProcess.java [hunk 2]
```diff
@@ -160,57 +141,72 @@ public void run() {
...
       }
-      LOG.log(Level.FINEST, "Process \"{0}\" returned {1}", new Object[]{this.id, returnValue});
+
+      LOG.log(Level.FINER, "Process \"{0}\" returned {1}", new Object[] {this.id, returnValue});
+
     } catch (final InterruptedException ex) {
-      LOG.log(Level.SEVERE, "Interrupted while waiting for the process \"{0}\" to complete. Exception: {2}",
-          new Object[]{this.id, ex});
+      LOG.log(Level.SEVERE,
+          "Interrupted while waiting for the process \"{0}\" to complete. Exception: {2}",
```
Comment (tcNickolas) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/process/RunnableProcess.java]: `Exception: {1}`?

Comment (motus) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/process/RunnableProcess.java]: Good catch! Sometimes even simple code reformatting helps to find bugs :)

**P.S.** I am done with the fixes. @tcNickolas could you please take another look? Thank you!

**Comment by tcNickolas** [comment]
LGTM and passes tests, I'll merge

**Pull request closed without merging**